### DISCUSSION TITLE
testing ice/turn without success

### DISCUSSION THREAD

**Discussion by juha-h** [baresip/baresip#2232] [Q&A] [answered]
I configured in my account:
```
;stunserver=turn:openrelay.metered.ca:80;stunuser=openrelayproject;stunpass=openrelayproject;medianat=ice
```
Then I tried to make a call:
```
test@sip.tutpro.com: selected for request
call uri: sip:foo@sip.tutpro.com
call: alloc with params laddr=192.168.238.215, af=AF_INET, use_rtp=1
ice: new session with TURN-server at openrelay.metered.ca (username=openrelayproject)
call: use_video=0
call: connecting to 'sip:foo@sip.tutpro.com'..
presence: ua=0x55ca102f7c10 got event 11 (CALL_OUTGOING)
call id: dc48c4d6cca68f0f
ice: resolved TURN-server to address 216.39.253.10:80
```
After that nothing happened.

I tested on page https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/ that the turn server is operational.

Any idea why nothing happens?  I started baresip with -v, but looks like there is no ice/turn related debug messages.

Could someone else try the same? Does it work for you?

My host is connected to mobile WiFi hotspot.

**Comment by juha-h** [comment]
I checked with wireshark and no packets were sent to 216.39.253.10:80 during the call attempt.  Also, the call attempt just hangs there and when I after a couple of minutes pressed ESC, the response was "no active call".

It starts to feel like this needs to be turned into a new issue.

**Comment by juha-h** [comment]
I tried with google stun server like this:
```
stunserver=stun:stun.l.google.com:19302;medianat=ice
```
When I dial, I get:
```
>                              bar
test@sip.tutpro.com: selected for request
call uri: bar
ice: new session with STUN-server at stun.l.google.com (username=)
call: connecting to 'sip:bar@sip.tutpro.com'..
call id: 9eeeaadc0eebd1c3
```
and after that nothing.  Wireshark doesn't show any traffic to port 19302.

What is it that I'm missing here? And if I have done something wrong, why don't I get any error message from baresip?

**Comment by cspiel1** [comment]
Do you have all the information already in the issue https://github.com/baresip/baresip/issues/2235? Then we could delete the discussion, right?

**Comment by cspiel1** [comment]
What is the complete account? I don't understand the your setup. Do have a registrar, or a sip proxy?

**Comment by juha-h** [comment]
Nothing special. Yes, registrar and outbound proxy.
```
<sip:test@sip.tutpro.com>;auth_user=test;auth_pass="XXX";outbound="sip:X.X.X.X:5061;transport=tls";ptime=20;audio_codecs=;audio_codecs=AMR-WB/16000/1,AMR/8000/1,opus/16000/1,PCMU,PCMA;video_codecs="";regint=600;pubint=0;regq=0.5;sipnat=outbound;call_transfer=yes;mwi=no;stunserver=stun:@stun.l.google.com:19302;medianat=ice
```

**Reply by cspiel1** [reply]
Maybe I should also setup such a configuration. Which outbound proxy and which SIP registrar do you use in your setup?

**Comment by juha-h** [comment]
I have kamailio /rtpengine/sems based SIP server platform. If you want, I can setup an account for you.

**Reply by cspiel1** [reply]
Yes, please this would help. I have too reproduce this. Will try this tomorrow.

**Reply by juha-h** [reply]
Send me email jh@tutpro.com and I'll give you the details.

**Comment by cspiel1** [comment]
- `call_streams_alloc()` should be called in `sipsess_desc_handler()`. And this is called during `send_invite()`.
- `call_streams_alloc()` allocates audio/video streams and `stream_alloc()` calls `mnat->mediah()`.
- The `mnat_connected_handler()` should be called asynchronously. After INVITE was sent already. Is this a problem?

On the other hand,
- In `call_alloc()` for the mnat case `mnat->sessh()` is called.
- The `mnat_handler()` calls `send_invite()`

Thus if I am correct, in this commit e80b7863ddabe9fda93845738a4760525b4f2676 the call to `mnat->mediah` was postponed to the response handler of the `mnat->sessh`.

Didn't you see this log message from function `mnat_connected_handler()`?
```c
	info("stream: '%s' mnat '%s' connected: raddr %J %J\n",
	     media_name(strm->type),
	     strm->mnat->id, raddr1, raddr2);
```

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Didn't you see this log message from function `mnat_connected_handler()`?
> ```c
> 	info("stream: '%s' mnat '%s' connected: raddr %J %J\n",
> 	     media_name(strm->type),
> 	     strm->mnat->id, raddr1, raddr2);
> ```

No, with current main this is all that shows up and then nothing:
```
>                              bar
***@***.***: selected for request
call uri: ***@***.***
ice: new session with STUN-server at stun.l.google.com (username=)
call: connecting to ***@***.***'..
call id: f4f4545b5d1e2dc1
```

**Comment by juha-h** [comment]
Christian Spielberger writes:

> - `call_streams_alloc()` should be called in `sipsess_desc_handler()`. And this is called during `send_invite()`.
> - `call_streams_alloc()` allocates audio/video streams and `stream_alloc()` calls `mnat->mediah()`.
> - The `mnat_connected_handler()` should be called asynchronously. After INVITE was sent already. Is this a problem?
> 
> On the other hand,
> - In `call_alloc()` for the mnat case `mnat->sessh()` is called.
> - The `mnat_handler()` calls `send_invite()`
> 
> Thus if I am correct, in this commit
> e80b7863ddabe9fda93845738a4760525b4f2676 the call to `mnat->mediah`
> was postponed to the response handler of the `mnat->sessh`.

With current main, mediah is not called at all.  Baresip at 2022-09-26
23:59 calls it, but not after that.

**Comment by juha-h** [comment]
Just to clarify since there is two media handlers (menc and mnat).  It is the latter that I meant.  So mnat sessh is called, but mediah is not.

**Comment by sreimers** [comment, marked as answer]
I could reproduce the same issue with TURN only. Looks like every mnat call_connect handling is broken. This seems to fix it at least for TURN:

```diff
diff --git a/src/call.c b/src/call.c
index 44e8144b..bb6c2467 100644
--- a/src/call.c
+++ b/src/call.c
@@ -1140,6 +1140,8 @@ int call_connect(struct call *call, const struct pl *paddr)
         * wait until completed before sending the INVITE */
        if (!call->acc->mnat)
                err = send_invite(call);
+       else
+               call_streams_alloc(call);
 
        return err;
 }
```

**Comment by juha-h** [comment]
Yes, I tested with that diff and call with medianat=turn worked BUT only when I build baresip binary with make command and start baresip in baresip source dir with ./baresip command. It does not work if I build baresip Debian package, install it and then start baresip with baresip command.  This is weird.  While wondering this, I noticed that `make distclean` does not clean obj-x86_64-linux-gnu directory.

**Reply by sreimers** [reply]
You have to install/update the libbaresip deb package, the baresip binary is linked with libbaresip.so.

**Comment by juha-h** [comment]
Yes, just figured that out.  Earlier (if I remember correctly) it was enough to install baresip package.  Now baresip depends on libbaresip and I had old one hanging around.

Once I installed fresh baresip and libbaresip packages, call with medianat=turn worked OK with your patch. I'll do more tests.

**Comment by juha-h** [comment]
I also tested with stun and ice.  With the patch, both worked.

**Comment by cspiel1** [comment]
The commit looks okay. I guess that the SDP `sess->laddr` is set by the `mnat` somehow. Hopefully `laddr` is correct.

This is the fallback `laddr` in ua.c:
```c
laddr = ua_regladdr(ua);
```

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Do you have all the information already in the issue
> https://github.com/baresip/baresip/issues/2235? Then we could delete
> the discussion, right?

Right, there is enough info in the issue.
